Design/Fit functions in python.md

```diff
@@ -0,0 +1,179 @@
+
+## Function construction
+
+For each concrete function type (Gaussian, Lorentzian, etc) there should exist a corresponding python function with the same name automatically generated in a way similar to the algorithm functions. A function object is then constructed by calling such a function (constructor) and is an instance of `FitFunctionWrapper`. For example
+```
+  g = Gaussian()
+```
+A call without passing in any initialisation arguments should create a function with default parameters and attributes.
+Initial values of parameters and attributes can be given via keyword arguments, for example
+```
+  p = Polynomial(attributes={'n': 3}, parameters={'A0': 1, 'A1': 2, 'A3': 3, 'A4': 4})
+```
+Parameter and attribute names can also be used as the keys:
+```
+  g = Gaussian(Height=1, Sigma=0.1)
+  
+  p = Polynomial(n=3, A0=1, A1=2, A3=3, A4=4)
+```
+The implementation must check each keyword argument if it's a parameter or an attribute by calling `IFunction::hasAttribute(name)`. All passed attributes must be set before setting parameters. There is need to be a provision for setting attributes in a particular order in case it is essential for a fitting function. This can be done for example by passing a list of name/value pairs (tuples or lists) to the `attributes` keyword argument.
+
+### Construction of composite functions
+
+Composite functions can be created in a similar way, by callig a constructor. It should be able to accept the same types of arguments as a simple function but also it must have the `functions` keyword argument that takes a list of member functions. For example:
+```
+  c = CompositeFunction(functions=[Gaussian(PeakCentre=1), Gaussian(PeakCentre=2)])
+```
+Another way of setting member functions is via positional arguments of the condtructor:
+```
+  c = CompositeFunction(Gaussian(PeakCentre=1), Gaussian(PeakCentre=2))
+```
+`FitFunctionWrapper` class will override the `__add__()` method to allow construction of composite functions from a sum of member functions:
+```
+  c = LinearBackground() + Gaussian(PeakCentre=1) + Gaussian(PeakCentre=2)
+```
+`ProductFunction` can be constructed from a product of member functions:
+```
+  p = ExpDecay() * UserFunction(Formula='sin(w*x)', w=1)
+```
+
+### Multi-domain functions
+
+A [`MultiDomainFunction`](https://github.com/mantidproject/mantid/blob/master/Framework/API/inc/MantidAPI/MultiDomainFunction.h) needs a custom constructor which sets domain indices to its members.
+
+## Updating functions
+
+### Getting and setting parameters and attributes
+
+`FitFunctionWrapper` will implement `__getitem__(self, name)` and `__setitem__(self, name, value)` methods to access parameters and attributes in a dictionary-like style (`name` is a string):
+```
+  sigma = gaussian['Sigma']
+  
+  comp_func['f1.A0'] = 0.5
```

Comment:
`comp_func[1]['A0'] = 0.5` will work because `comp_func[1]` returns a function. But `'f1.A0'` is a valid parameter name and the function must be able to set a value for it.
`comp_func[1].A0 = 0.5` will also make perfect sense.